ui/app/ducks/gas/gas.duck.js

```diff
@@ -451,6 +463,52 @@ export function fetchGasEstimates (blockTime) {
   }
 }
 
+export function fetchMetaSwapsGasPriceEstimates () {
+  return async (dispatch, getState) => {
+    const { basicPriceAndTimeEstimatesLastRetrieved } = getState().gas
+    const timeLastRetrieved = basicPriceAndTimeEstimatesLastRetrieved || loadLocalStorageData('METASWAP_GAS_PRICE_ESTIMATES_LAST_RETRIEVED') || 0
+
+    dispatch(basicGasEstimatesLoadingStarted())
+
+    let basicEstimates
+    if (Date.now() - timeLastRetrieved > 30000) {
+      basicEstimates = await fetchExternalMetaSwapGasPriceEstimates(dispatch)
+    } else {
+      const cachedBasicEstimates = loadLocalStorageData('METASWAP_GAS_PRICE_ESTIMATES')
+      basicEstimates = cachedBasicEstimates || await fetchExternalMetaSwapGasPriceEstimates(dispatch)
+    }
+
+    dispatch(setBasicGasEstimateData(basicEstimates))
+    dispatch(basicGasEstimatesLoadingFinished())
+    return basicEstimates
+  }
+}
+
+async function fetchExternalMetaSwapGasPriceEstimates (dispatch) {
+  const response = await queryMetaSwapsGasPriceApi()
+
+  const {
+    SafeGasPrice: safeLow,
+    ProposeGasPrice: average,
+    FastGasPrice: fast,
+    LastBlock: blockNum,
+  } = await response.json()
```

Comment:
It seems a bit strange to me that this isn't done in the `queryMetaSwapsGasPriceApi` function. A fetch response isn't the most intuitive return value - nothing can use it without calling `.json()` first and checking the return status.

Speaking of which - the return status isn't being checked, so this object might be a description of an error. Minimally we should be ensuring `response.ok` is truthy before this step.

Reply:
Also we should validate these responses. Perhaps also in the `queryMetaSwapsGasPriceApi` function?

Reply:
the first issue mentioned here went away with the refactor and validation was done in bbc7f52e4